lib/server/routes.js

```diff
@@ -64,6 +53,40 @@ router.get('*', (req, res, next) => {
       res.status(404).send('Not found')
     }
   })
+}
+
+router.all('/data/*', (req, res) => {
+  var splat = req.params[0]
+  if (splat == null) {
+    error('noope', res)
+  } else {
+    var url = req.url.replace(/^\/data\//, '')
+    routeWebApp(url, res, (renderProps, store) =>
+      res.status(200).send(views.data({
+        data: JSON.stringify(store.getState())
+      })))
+  }
+})
+
+router.get('*', (req, res) => {
+  routeWebApp(req.url, res, (renderProps, store) => {
+    const html = renderToString(
+      <Provider store={store}>
+        <RoutingContext {...renderProps}/>
+      </Provider>
+    )
+    const state = JSON.stringify(store.getState())
+    res.status(200).send(
+      views.layout({
+        title: (renderProps.params.title || '') + ' - Wikipedia',
+        body: html,
+        url: req.url,
+        // Send inline data when it is smaller than 30kb. Else let the browser
+        // request it with a script tag
+        data: state.length < 30 * 1024 ? state : undefined
```

Comment:
It'd be nice to trigger this with a query parameter but this'll do for now. Also, 30 KB is going to be larger than most lead sections, so this – rather nice (!) – optimisation won't be invoked for most initial requests.